qos/evm/state.go

```diff
@@ -127,90 +95,20 @@ func (s *ServiceState) UpdateFromEndpoints(updatedEndpoints map[protocol.Endpoin
 		// Update the perceived block number.
 		s.perceivedBlockNumber = blockNumber
 
-		// Attempt to retrieve the archival balance from the endpoint.
-		balance, err := endpoint.getArchivalBalance()
-		if err != nil {
-			logger.Info().Err(err).Msg("Skipping endpoint without archival balance")
+		// Process archival data from the endpoint (if enabled)
+		// This contributes to consensus regarding the expected archival balance
+		// at a specific historical block number
+		if !s.archivalState.processEndpointArchivalData(endpoint) {
+			logger.Info().Msg("Skipping endpoint without archival balance")
 			continue
 		}
-
-		// Only count non-empty balances toward consensus.
-		if s.archivalState.balance == "" && balance != "" {
-			s.archivalState.balanceConsensus[balance]++
-		}
 	}
 
-	// Update archival block number and archival balance only if not yet set.
-	if s.perceivedBlockNumber != 0 && s.archivalState.blockNumberHex == "" {
-		s.assignArchivalBlockNumber()
-	}
-	if s.archivalState.balance == "" {
-		s.updateArchivalBalance(archivalConsensusThreshold)
-	}
+	// Update the archival state based on the perceived block number
+	// This handles:
+	// 1. Calculating a random archival block number if needed
+	// 2. Determining consensus on the expected balance at that block number
+	s.archivalState.updateArchivalState(s.perceivedBlockNumber)
```

Comment:
Not a blocker, as I am not 100% sure it is possible: but making a single call to the archivalState and passing it all data that is relevant will make the code much easier to read.

Reply:
This is done.

See commit:
https://github.com/buildwithgrove/path/pull/194/commits/a4f04e376fbbe2c42bc4169ec9e4f8e900c0fd9d